### ISSUE TITLE
SSL routines:ssl3_get_record:decryption failed or bad record mac (openssl 3.0 git)

### ISSUE THREAD

**Issue by pawlosck** [openssl/openssl#11727]
Bug found in latest openssl version (3.0) (latest commit: today 2020-05-04)

I tried to upload 2 files using nextcloud and in apache logs I see these logs:
If I remember well, the same problem is in ubuntu 19.10.
++
SSL library error 1 reading data
SSL Library Error: error:1408F119:SSL routines:ssl3_get_record:decryption failed or bad record mac
++

Apache from latest ubuntu 20.04.

Options used during build openssl
./config --prefix=/usr/local/ssl --openssldir=/usr/local/ssl shared zlib
make
make test -> no errors

EDIT:
cert in apache: self signed

**Comment by mattcaswell** [comment]
> If I remember well, the same problem is in ubuntu 19.10.

Are you saying this same issue exists with a stable version of OpenSSL in ubuntu 19.10? i.e. this is not a regression in 3.0?

Is this issue reproducible? Do I understand correctly that OpenSSL is being used as the server here? What is the client?

**Comment by pawlosck** [comment]
> > If I remember well, the same problem is in ubuntu 19.10.
> 
> Are you saying this same issue exists with a stable version of OpenSSL in ubuntu 19.10? i.e. this is not a regression in 3.0?

I don't remember what version it was, because I tested 2 or 3 stable versions and dev version. Generally I have a lot problems with https in apache (openssl stable and dev).
For example, yesterday I found next bug connected with OpenSSL (probably). Bug found when I tried to use android nextcloud client.
++
(104)Connection reset by peer: [client IP:PORT] AH01991: SSL input filter read failed.
++
It happen when I tried replay audio file from android nextcloud client.

I had problems with handshake too when I tried to use TLS1.3 in apache.

> 
> Is this issue reproducible? Do I understand correctly that OpenSSL is being used as the server here? What is the client?

Yes, this issue is reproducible 100/100.

I don't know how to answer. I don't use OpenSSL directly (except preparing self signed cert). I configured apache to use https connection and apache probably use openssl via ssl module.

Currently, client is Firefox for ubuntu (version 75) -> desktop

**Comment by mattcaswell** [comment]
The "decryption failed or bad record mac" error can be caused by a wide variety of reasons. Typically this is due to the data being corrupted by the time it gets to the record decryption in OpenSSL - whether that be because of some server bug (whether in OpenSSL or some higher application layer), network middlebox interference, or client side issues. Sometimes it can be helpful to get a wireshark trace of a failed connection. I'd suggest that as a next step.

**Comment by pawlosck** [comment]
I connected directly (via cable) computer with server and except one error (other than "decryption failed or bad record mac", but still connected with openssl), and I didn't see any problems, so probably bug is connected with router. I don't have other router to verify it, but possible in few days I will have access to other router and then I will verify it.

Could you tell me what I should check using wireshark (protocols, fields in protocols etc)? If possible, send me wireshark filter.

**Comment by mattcaswell** [comment]
Assuming the server is listening on port 443, just using the "ssl" display filter should be sufficient.

**Comment by pawlosck** [comment]
I am still checking with new router, but probably I know where is problem.
It happen always, when I am trying to use public ip to connect to nextcloud from LAN(LAN_IP_PC -> PUBLIC_IP_SERVER ).
When I am using LAN IP of server to connect to nextcloud, then everythong is ok.
In router there is forwarding. When I am using Public ip with high port (let see PUBLIC_IP:44444), then router forwarding traffic to LAN IP and https port (for example: 192.168.1.222:443).

**Comment by kroeckx** [comment]
@pawlosck: Can this issue be closed?

**Comment by pawlosck** [comment]
Yes. I prepared cert from let's encrypt and all problems fixed. I don't know where is problem. Probably, some problems are connected with Netgear's routers, but other problems were on other routers too.
I think, some problems can be connected with forwarding on Netgear's routers. If I rermember well, Negear's routers wrongly translate ip when trying to connect to external IP(domain) from LAN. I checked routers from two other companies (other than Netgear) and main problem fixed, but not all. Maybe ssl on router do something wrong. I don't know.
Self signed certs are great, but currently a lot of apps don't want to accept it.

**Comment by FarhanAliRaza** [comment]
> Yes. I prepared cert from let's encrypt and all problems fixed. I don't know where is problem. Probably, some problems are connected with Netgear's routers, but other problems were on other routers too. I think, some problems can be connected with forwarding on Netgear's routers. If I rermember well, Negear's routers wrongly translate ip when trying to connect to external IP(domain) from LAN. I checked routers from two other companies (other than Netgear) and main problem fixed, but not all. Maybe ssl on router do something wrong. I don't know. Self signed certs are great, but currently a lot of apps don't want to accept it.

Please how can you tell how you generated certificate as i am getting this error on my local machine

**Comment by pawlosck** [comment]
I don't remember, but you can see all ssl commands from 2020. Maybe it help you, but I am not sure is it the same commands which I used to prepare selfsigned cert.


openssl x509 -noout -text -in server.CA.crt                                                                                                                                                                   
openssl ca -out server.apache.pem -keyfile server.CA.key -infiles server.apache.csr
openssl ca -out server.apache.pem -keyfile ../ca/server.CA-signed.crt -infiles server.apache.csr                                                                                                              
openssl ca -out server.apache.pem -keyfile ../ca/server.CA.key  -infiles server.apache.csr                                                                                                                     
openssl genrsa -des3 -out server.CA.key 2048                 
openssl req -verbose -new -key server.CA.key -out server.CA.csr -sha256
openssl x509 -noout -text -in server.CA-signed.crt               
openssl ca -extensions v3_ca -out ../CA/server.CA-signed.crt  -keyfile ../CA/server.CA.key  -verbose -selfsign -md sha256 -enddate 301231235959Z -infiles server.CA.csr                                       
openssl ca -out server.apache.pem -keyfile ../CA/server.CA.key  -infiles server.apache.csr                                                                                             
openssl x509 -inform PEM -outform DER -in server.CA-signed.crt  -out server.CA-signed_format_der_android.crt                                                                                                   
openssl rsa -in server.apache.key -out server.apache.without_password.key                                                                                                             
openssl ca -extensions v3_ca -out server.CA-signed.crt -keyfile server.CA.key -verbose -selfsign -md sha256 -enddate 301231235959Z -infiles server.CA.csr                                                     
openssl rsa -in server.apache.key -out server.apache.unsecured.key                       
openssl genrsa -des3 -out server.apache.key 2048
openssl req -verbose -new -key server.apache.key -out server.apache.csr -sha256                   
openssl ca -out server.apache.pem -keyfile ../CA/server.CA.key -infiles server.apache.csr           
openssl rsa -in server.apache.key -out server.apache.key_without_password.pem

**Comment by pawlosck** [comment]
Are you using any Netgear hardware? I am almost sure, problem is related to Netgear routers.

**Comment by mattcaswell** [comment]
> Please how can you tell how you generated certificate as i am getting this error on my local machine

"decryption failed or bad record mac" is a very generic error which can be the result of a large number of different problems.

@FarhanAliRaza it is quite unlikely that your problem is the same one encountered by @pawlosck. If it is still a problem for you I'd recommend creating a new issue.

**Comment by geektcp** [comment]
I see the same issue, but I have no solution to fix it. I suspect this issue is related to OpenSSL.
```
yarn install v1.22.19
info No lockfile found.
[1/4] Resolving packages...
[2/4] Fetching packages...
error An unexpected error occurred:
 "https://registry.npmmirror.com/regenerator-runtime/-/regenerator-runtime-0.14.0.tgz: incorrect data check".
info If you think this is a bug, 
please open a bug report with the information provided in "/code/test/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install 
for documentation about this command.
```

- os
```
CentOS7.9
```

- node
```
node --version
v16.20.2

npm --version
8.19.4

yarn --version
1.22.19

openssl version
OpenSSL 1.0.2k-fips  26 Jan 2017
```

**Comment by acudworth3** [comment]
seeing this on wslw + ubuntu  as well. Might be an issue with wsl, but wget curl etc mostly seem to work.

**Comment by cahyosubroto** [comment]
Hi so what is the resolution for this?

**Comment by ky0n** [comment]
i got this on wsl + arch linux on 2 pcs, any hints?

**Comment by acudworth3** [comment]
> i got this on wsl + arch linux on 2 pcs, any hints?

This might sound crazy, but I had a docking station that I think somehow was interfering with the decryption. Maybe its coincidence, but I disconnected the hub and all monitors. Then retried and it started working. I have updated WSL a few times since then and never run into the issue. Once or twice I'll get this error when doing HTTPS and removing the hub resolves it. I'd triple check there isn't any hardware between your WSL and the internet that could be getting in the way unexpectedly.
